tiatoolbox/dataloader/wsireader.py

```diff
@@ -84,15 +85,148 @@ def slide_info(self):
+            else:
+                raise ValueError("MPP is None")
+        elif units == "power":
+            if info.objective_power is not None:
+                base_scale = 1 / info.objective_power
+                target_scale = 1 / target_scale
+            else:
+                raise ValueError("Objective power is None")
+        elif units == "level":
+            if target_scale >= len(info.level_downsamples):
+                raise ValueError("Target scale level > number of levels in WSI")
+            base_scale = 1
+            target_scale = info.level_downsamples[target_scale]
+        elif units == "baseline":
+            base_scale = 1
+            target_scale = 1 / target_scale
+        else:
+            raise ValueError("Invalid units")
+
+        return [(base_scale * ds) / target_scale for ds in info.level_downsamples]
+
+    def optimal_relative_level_scale(self, target_scale, units, precision=3):
+        """
+        Find the optimal level to read at for a desired scale and units.
+
+        The optimal level is the most downscaled level of the image
+        pyramid (or multi-resolution layer) which is larger than the
+        desired target scale. The returned scale is the scale factor
+        required, post read, to achieve the desired scale.
+
+        Args:
+            target_scale (float): Scale to calculate relative to
+            units (str): Units of the scale. Allowed values are the same
+                as for WSIReader.relative_level_scales
+            precision (int, optional): Decimal places to use when
+                finding optimal scale. This can be adjusted to avoid
+                errors when an unecessary precision is used. E.g.
+                1.1e-10 > 1 is insignificant in most cases.
+                Defaults to 3.
+
+        Returns:
+            tuple: Optimal read level and scale factor between the optimal level
+                and the target scale (usually <= 1).
+        """
+        level_scales = self.relative_level_scales(target_scale, units)
+        # Note that np.argmax finds the index of the first True element.
+        # Here it is used on a reversed list to find the first
+        # element <=1, which is the same element as the last <=1
+        # element when counting forward in the regular list.
+        reverse_index = np.argmax(
+            [np.all(np.round(x, decimals=precision) <= 1) for x in level_scales[::-1]]
+        )
+        # Convert the index from the reversed list to the regular index (level)
+        level = (len(level_scales) - 1) - reverse_index
+        scale = level_scales[level]
+        if np.any(np.array(scale) > 1):
+            warnings.warn(
+                "Scale > 1."
+                "This means that the desired scale is a higher"
+                " resolution than the WSI can produce."
+                "Interpolation of read regions may occur."
+            )
+        return level, scale
+
+    def read_rect_params_for_scale(
```

Comment:
read_params_for_target_scale() ?

Reply:
During refactoring I renamed to `find_read_rect_params` before I remembered your comment. Is that any better in you opinion. Still open to suggestions. Maube `find_optimal_read_rect_params`? Also considering making 'private' i.e. `_find_read_rect_params`.